presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/GroupedTypedHistogram.java

```diff
@@ -0,0 +1,552 @@
+            nextNodePointer++;
+        }
+    }
+
+    private boolean isCurrentGroupEmpty()
+    {
+        return headPointers.get(currentGroupId) == NULL;
+    }
+
+    private void iterateGroupNodes(long groupdId, NodeReader nodeReader)
+    {
+        int currentPointer = (int) headPointers.get(groupdId);
+
+        if (currentPointer == NULL) {
+            nodeReader.read(NULL);
+        }
+        else {
+            while (currentPointer != NULL) {
+                checkState(currentPointer < nextNodePointer, "error, corrupt pointer; max valid %s, found %s", nextNodePointer, currentPointer);
+                nodeReader.read(currentPointer);
+                currentPointer = nextPointers.get(currentPointer);
+            }
+        }
+    }
+
+    private void rehash()
+    {
+        long newBucketCountLong = bucketCount * 2L;
+
+        if (newBucketCountLong > Integer.MAX_VALUE) {
+            throw new PrestoException(GENERIC_INSUFFICIENT_RESOURCES, "Size of hash table cannot exceed 1 billion entries");
+        }
+
+        int newBucketCount = (int) newBucketCountLong;
+        int newMask = newBucketCount - 1;
+        IntBigArray newBuckets = new IntBigArray(-1);
+        newBuckets.ensureCapacity(newBucketCount);
+
+        for (int i = 0; i < nextNodePointer; i++) {
+            // find the old one
+            int bucketId = getBucketIdForNode(i, newMask);
+            int probeCount = 1;
+
+            int originalBucket = bucketId;
+            // find new one
+            while (newBuckets.get(bucketId) != -1) {
+                int probe = nextProbe(probeCount);
+                bucketId = nextBucketId(originalBucket, newMask, probe);
+                probeCount++;
+            }
+
+            // record the mapping
+            newBuckets.set(bucketId, i);
+        }
+
+        buckets = newBuckets;
+        bucketCount = newBucketCount;
+        maxFill = calculateMaxFill(newBucketCount, MAX_FILL_RATIO);
+        mask = newMask;
+        resizeNodeArrays(newBucketCount);
+    }
+
+    private int nextProbe(int probeCount)
+    {
+        switch (PROBE_TYPE) {
+            case LINEAR:
+                return nextProbeLinear(probeCount);
+            case SUM_OF_COUNT:
+                return nextSumOfCount(probeCount);
+            case SUM_OF_SQUARE:
+                return nextSumOfSquares(probeCount);
+        }
+
+        throw new IllegalArgumentException(String.format("invalid enum %s", PROBE_TYPE));
+    }
+
+    //parallel arrays with data for
+    private void resizeNodeArrays(int newBucketCount)
+    {
+        // every per-bucket array needs to be updated
+        counts.ensureCapacity(newBucketCount);
+        valuePositions.ensureCapacity(newBucketCount);
+        nextPointers.ensureCapacity(newBucketCount);
+        groupIds.ensureCapacity(newBucketCount);
+        valueRawHashList.ensureCapacity(newBucketCount);
+    }
+
+    private long getHashPosition(long groupId, long valueRawHash)
```

Comment:
This method is mis-named. It returns a hash, not a position.

Reply:
yea, holdover from old code i think.

I'll rename to just "getValueRawHash()" since there are multiple hashes we deal with

Reply:
actually I'll rename this method and var to "groupAndValueHash"

in reality what i've done is broken down the value hash to one by value only and one with group since we store a hash of values per-group. i'lll make a pass and see if i can make the names of these hash variables and functions cleaner

Reply:
renamed